Expand Number to support the full integer range

ximinez opened the pull request:
Supersedes #6000.

- Update the conversion points between Number and {XRP,MPT,IOU,ST}Amount & STNumber.

<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

Bronek commented on include/xrpl/basics/Number.h:
```diff
@@ -304,29 +428,41 @@ operator/(Number const& x, Number const& y)
 Number::lowest() noexcept
 {
-    return -Number{maxMantissa, maxExponent, unchecked{}};
+    return -Number{range_.get().max, maxExponent, unchecked{}};
```
this function is no longer `constexpr` because it reads from `static thread_local range_`

Bronek commented on include/xrpl/basics/Number.h:
```diff
@@ -304,29 +428,41 @@ operator/(Number const& x, Number const& y)
 Number::max() noexcept
 {
-    return Number{maxMantissa, maxExponent, unchecked{}};
+    return Number{range_.get().max, maxExponent, unchecked{}};
```
this function is no longer `constexpr` because it reads from `static thread_local range_`

Bronek commented on include/xrpl/basics/Number.h:
```diff
@@ -304,29 +428,41 @@ operator/(Number const& x, Number const& y)
 Number::min() noexcept
 {
-    return Number{minMantissa, minExponent, unchecked{}};
+    return Number{range_.get().min, minExponent, unchecked{}};
```
this function is no longer `constexpr` because it reads from `static thread_local range_`

Bronek commented on include/xrpl/basics/Number.h:
```diff
@@ -304,29 +428,41 @@ operator/(Number const& x, Number const& y)
 inline constexpr bool
 Number::isnormal() const noexcept
 {
+    MantissaRange const& range = range_;
```
this function is no longer `constexpr` because it reads from `static thread_local range_`

Bronek reviewed (commented):
Pls fix compilation errors because several functions are no longer `constexpr`

Bronek commented on src/libxrpl/basics/Number.cpp:
This assert fails in AMM unit tests

```
ripple.app.AMM Basic Payment
rippled: ../src/libxrpl/basics/Number.cpp:451: ripple::Number& ripple::Number::operator*=(const ripple::Number&): Assertion `("ripple::Number::operator*=(Number) : is normal") && (isnormal() && y.isnormal())' failed.
[1]    364083 IOT instruction (core dumped)  ./rippled -u AMM
```

ximinez replied:
> This assert fails in AMM unit tests
> 
> ```
> ripple.app.AMM Basic Payment
> rippled: ../src/libxrpl/basics/Number.cpp:451: ripple::Number& ripple::Number::operator*=(const ripple::Number&): Assertion `("ripple::Number::operator*=(Number) : is normal") && (isnormal() && y.isnormal())' failed.
> [1]    364083 IOT instruction (core dumped)  ./rippled -u AMM
> ```

Though there are still some failing tests at the moment, no asserts are failing anymore.

gregtatcam commented on include/xrpl/basics/Number.h:
```diff
@@ -14,41 +17,97 @@ std::string
+    int log;
+    mantissa_scale scale;
+};
+
```
This PR modifies `Number` behavior. Please add comments describing what this new behavior solves, how it solves it, and if there are any changes to how `Number` should be used. This is a core class used in all calculations and a detailed description will help with review and future maintenance.

ximinez replied:
> This PR modifies `Number` behavior. Please add comments describing what this new behavior solves, how it solves it, and if there are any changes to how `Number` should be used. This is a core class used in all calculations and a detailed description will help with review and future maintenance.

Added in 16ab78d3dfe828c597aeddfa19c70f0368762285

ximinez replied:
> > This PR modifies `Number` behavior. Please add comments describing what this new behavior solves, how it solves it, and if there are any changes to how `Number` should be used. This is a core class used in all calculations and a detailed description will help with review and future maintenance.
> 
> Added in [16ab78d](https://github.com/XRPLF/rippled/commit/16ab78d3dfe828c597aeddfa19c70f0368762285)

Just for completeness: Also 6303dd5314e71ef898fa7728b1d8c9f403f9bcd7.

gregtatcam commented on include/xrpl/protocol/STTakesAsset.h:
```diff
@@ -0,0 +1,29 @@
+#include <xrpl/protocol/STBase.h>
+
+namespace xrpl {
+
```
Please add comments describing this class.

shawnxie999 replied:
+1

What problem exactly does it solve?

gregtatcam replied:
Is this class even needed at all? Why not just add `optional<Asset>` and `associateAsset()` directly to `STNumber`?

gregtatcam replied:
Also, we don't really need to know an asset associated with the number, we just need to know if it's integral or not.  Then having integral flag on `STNumber` rather than inheriting from `STTakesAsset` would makes sense tool.

ximinez replied:
> Is this class even needed at all? Why not just add `optional<Asset>` and `associateAsset()` directly to `STNumber`?

By itself, no, the `STTakesAsset` class doesn't do much.  It just stores an `Asset` if it's given one. It's intended to be used as a base class for any `ST` type that needs to use an `Asset` when it doesn't have one intrinsically like `STIssue` and `STAmount` do.

Right now, that's only `STNumber`, but I wrote it to be a little bit future-proof so that if the functionality needs to be added in the future to a new type, or even an existing type, you can plug it in anywhere. Then in the derived class you can override `associateAsset`, and use `asset_` anywhere it's needed with no extra glue or boilerplate.

I could have written the handling directly into `STNumber` , but then if we ever wanted to reuse `sMD_NeedsAsset` on another type, we'd have to, at minimum, update the global `associateAsset` to handle the new type. This just seemed like it would save a step.

ximinez replied:
> Also, we don't really need to know an asset associated with the number, we just need to know if it's integral or not. Then having integral flag on `STNumber` rather than inheriting from `STTakesAsset` would makes sense tool.

The advantage of storing an asset is that that it lets me use the `roundToAsset` function to do the rounding. That particular overload just instantiates an `STAmount` with the asset and assigns that value back to the reference parameter. That takes advantage of `STAmount`'s internal rounding logic.

If I instead only had a flag, I would have to either reinvent the wheel - implement yet another rounding scheme - or create an asset of the correct type - XRP/MPT or IOU - whenever I want to do rounding.

gregtatcam replied:
Sure, but it's storing 40 bytes for Asset vs 1 byte for the flag. Ultimately any rounding takes mantissa, exponent, and integral flag. There could be a function that takes these three arguments and returns updated mantissa and exponent. This function can be overloaded for STAmount, etc. Just a thought.

ximinez replied:
> Sure, but it's storing 40 bytes for Asset vs 1 byte for the flag. Ultimately any rounding takes mantissa, exponent, and integral flag. There could be a function that takes these three arguments and returns updated mantissa and exponent. This function can be overloaded for STAmount, etc. Just a thought.

It's important to draw the distinction that the `Asset` in `STNumber` _is not stored on ledger_. It uses memory, but only for the duration of the transaction. We don't have to worry about storing umpteen millions of `Asset`s for an indefinite amount of time, but only a dozen or so for the time it takes to process a transaction.

shawnxie999 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -250,57 +283,168 @@ Number::normalize()
+    bool& negative,
+    unsigned long long& mantissa,
+    int& exponent,
+    internalrep const& minMantissa,
+    internalrep const& maxMantissa)
+{
```
can't we fetch the mantissa directly from `range_`, why do we need to have new params?

ximinez replied:
> can't we fetch the mantissa directly from `range_`, why do we need to have new params?

Because of `normalizeToRange`. That function is public and has template parameters. It calls `normalize` with min and max values, which can, in principal, be any valid value.

`normalizeToRange` is called from `STAmount::fromNumber` and `IOUAmount::fromNumber`. That is _how_ we convert from `Number` to those types, scaling the mantissa down if necessary, and very intentionally reuses the same code so that the same rounding rules are used regardless of the current scale.

shawnxie999 commented:
doesn't this need to be gated by an amendment?

gregtatcam commented on include/xrpl/basics/Number.h:
```diff
@@ -14,41 +17,97 @@ std::string
     explicit Number(rep mantissa, int exponent);
-    explicit constexpr Number(rep mantissa, int exponent, unchecked) noexcept;
+    explicit constexpr Number(
+        bool negative,
```
Suggestion - change the arguments order here and other constructors to be consistent with `STAmount`, which has `negative` after `exponent`.

ximinez replied:
> Suggestion - change the arguments order here and other constructors to be consistent with `STAmount`, which has `negative` after `exponent`.

I think I want to keep it this way because it more naturally represents the way a number is represented. e.g. `-23e34`: `[sign][mantissa][exponent]`.

AFAIK, the reason that the sign parameter is last in `STAmount` is that it defaults to `false` in _many_ of the ctor overloads, and those are used extensively throughout the code. Most `STAmount`s are positive, so it makes sense to make it easy to leave it off.

OTOH, the only `Number` ctors that allow you to even specify a sign flag are the "special" `unchecked` and `normalized` versions that also take an unsigned mantissa. Outside of `Number` itself, `unchecked` looks like it is only used in one place (an `STAmount` conversion), and `normalized` is only used in tests. I figure in tests where we're manually building numbers, I want to be as explicit as possible.

Very different circumstances and usage patterns.

shawnxie999 commented:
Want to make sure I understood this correctly: with largeRange, we can support full precision up to 2^63 - 1, so integer types (MPT or XRP) don’t lose precision even for large values. We’re still bounded by the signed `int64` limit, but this works for existing integers since none of them exceed `int64`. 

However, if we ever want to go beyond the 63-bit range (e.g., increasing MPT’s `MaximumAmount` to `max(uint64)` in the future), this implementation is not sufficient.

(Answered in this comment: https://github.com/XRPLF/rippled/pull/6025#issuecomment-3667384338)

gregtatcam commented on include/xrpl/basics/Number.h:
```diff
@@ -208,13 +401,34 @@ inline Number::Number(rep mantissa) : Number{mantissa, 0}
 Number::mantissa() const noexcept
 {
-    return mantissa_;
+    auto m = mantissa_;
```
I'm not following. This and next function are getters, which should just return the requested property as is and yet they might return a modified value. Please add comments explaining this change.

ximinez replied:
> I'm not following. This and next function are getters, which should just return the requested property as is and yet they might return a modified value. Please add comments explaining this change.

I explained this in the class documentation.

I added comments pointing to that explanation in 7265804170.

gregtatcam commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -345,10 +349,19 @@ STAmount::STAmount(
+    }
+    else
+    {
+        if (integral() && mValue > std::numeric_limits<std::int64_t>::max())
```
Shouldn't this be amendment gated?

shawnxie999 replied:
i believe it already is. The exception is in the `else` block, where `largeRange` is applied only if it’s set - which happens only when lending is enabled.

gregtatcam commented on include/xrpl/protocol/STAmount.h:
```diff
@@ -542,12 +555,31 @@ STAmount::operator=(XRPAmount const& amount)
+        std::uint64_t const intValue = static_cast<std::int64_t>(working);
+        return STAmount{asset, intValue, 0, negative};
+    }
+    else
```
`else` is redundant.

ximinez replied:
> `else` is redundant.

4605bbfb6f1e1464fb15b076a3ffcd7efe68d916

gregtatcam commented on include/xrpl/basics/Number.h:
```diff
@@ -299,30 +515,48 @@ operator/(Number const& x, Number const& y)
+}
+
+template <class T>
+std::pair<T, int>
```
Consider adding `require` to limit `T` to int64 and uint64.

ximinez replied:
> Consider adding `require` to limit `T` to int64 and uint64.

4605bbfb6f1e1464fb15b076a3ffcd7efe68d916

gregtatcam commented on include/xrpl/protocol/STNumber.h:
```diff
@@ -20,7 +21,7 @@ namespace xrpl {
  * that may be deduced from the context.
  */
-class STNumber : public STBase, public CountedObject<STNumber>
+class STNumber : public STTakesAsset, public CountedObject<STNumber>
```
Please add comments explaining why `STNumber` inherits from `STTakesAsset`.

ximinez replied:
> Please add comments explaining why `STNumber` inherits from `STTakesAsset`.

4605bbfb6f1e1464fb15b076a3ffcd7efe68d916

shawnxie999 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -387,101 +573,194 @@ divu10(uint128_t& u)
+        xe = zero.exponent_;
+        zn = zero.negative_;
+    }
+    if (xe > maxExponent)
```
why are we removing doRoundUp?

ximinez replied:
> why are we removing doRoundUp?

Laziness played a big part in it. These blocks conflicted when I merged changes up from `develop`, and it was easier to keep my changes than to move them to the `doRoundUp` and others, while making sure no information was lost. It might be worth recreating them.

ximinez replied:
> > why are we removing doRoundUp?
> 
> Laziness played a big part in it. These blocks conflicted when I merged changes up from `develop`, and it was easier to keep my changes than to move them to the `doRoundUp` and others, while making sure no information was lost. It might be worth recreating them.

Brought back in a06de68ce4

shawnxie999 replied:
thanks for adding it back!

shawnxie999 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -250,57 +283,168 @@ Number::normalize()
+    // the final mantissa_ is going to end up larger to fit within the range.
+    // Cut it down here so that the rounding will be done while it's smaller.
+    //
+    // Example: 9,900,000,000,000,123,456 > 9,223,372,036,854,775,808,
```
```suggestion
    // Example: 9,900,000,000,000,123,456 > 9,223,372,036,854,775,807,
```

the largest int64 is 9,223,372,036,854,775,80**7**

ximinez replied:
Fixed in a06de68ce4

shawnxie999 commented on src/libxrpl/basics/Number.cpp:
```diff
@@ -250,57 +283,168 @@ Number::normalize()
     }
+    // When using the largeRange, "m" needs fit within an int64, even if
+    // the final mantissa_ is going to end up larger to fit within the range.
+    // Cut it down here so that the rounding will be done while it's smaller.
```
could we elaborate? why does it have to be within int64 (during normalization)?  Wouldn’t we want to preserve as much precision as possible for as long as we can?

ximinez replied:
> could we elaborate? why does it have to be within int64 (during normalization)? Wouldn’t we want to preserve as much precision as possible for as long as we can?

An argument could be made either way. The main reason I did it this way is because `Number` is still conceptually an `int64`. If `Number` stored extra data, then it would have to handle rounding when truncating back down to `int64`, specifically in `mantissa()`. I also think it could introduce rounding errors because you'd effectively be rounding _twice_.

Example: Say you have a value like 999,900,000,000,000,005,455, and we're rounding to nearest. That's much bigger than `maxMantissa`. To get it under `maxMantissa`, we have to remove two digits, but to get it under `maxRep`, we need to remove three digits.
1. If we do it the current way, we divide it down to 999,900,000,000,000,005 with 0.455 as remainder. Since the remainder is < 0.5, we round down, so it stays 999,900,000,000,000,005.
2. If we preserve as much precision as possible, we only divide it down to 9,999,000,000,000,000,054, with 0.55 remainder. Since the remainder is > 0.5, we round up, resulting in 9,999,000,000,000,000,055.
	1. Now when `mantissa()` is called, if we round, the number is going to round up again to 999,900,000,000,000,006. It's obviously a tiny difference, but I think the _more_ correct answer is the value ending in 5.
	2. Alternatively we could truncate in `mantissa()`, but that would cause problems if the rounding was supposed to go the other way, such as with 999,900,000,000,000,005,744.

shawnxie999 replied:
Thanks for the detailed explanation! I agree the current implementation makes more sense.

shawnxie999 commented on include/xrpl/basics/Number.h:
```diff
@@ -177,28 +246,152 @@ class Number
+        MantissaRange::large,
+        1'000'000'000'000'000'000LL};
+    static_assert(isPowerOfTen(largeRange.min));
+    static_assert(largeRange.max == internalrep(9'999'999'999'999'999'999ULL));
```
What are the implications of converting a `Number` (with `largeRange`) to `STAmount` for an IOU amount? Since IOUs don’t support the full precision of `largeRange`, could this affect accounting, or is it fine because IOUs use floating-point representation which are already non-accurate anyways?

ximinez replied:
> What are the implications of converting a `Number` (with `largeRange`) to `STAmount` for an IOU amount? Since IOUs don’t support the full precision of `largeRange`, could this affect accounting, or is it fine because IOUs use floating-point representation which are already non-accurate anyways?

Somewhere else in this PR, I mentioned the `normalizeToRange` and `fromNumber` functions, which take the `Number` and rescale it, rounding as appropriate.

I realize that given my previous comment (https://github.com/XRPLF/rippled/pull/6025#discussion_r2670431755), this may sound like the same "rounding twice" problem. The difference is that when converting from a large `uint64` to `int64`, you're only going to drop one digit, giving that digit outsized influence on the rounding. When converting from a large `int64` to fit in an `STAmount`, you're going to be dropping at least _two_ digits. The first digit after the decimal point is the most important - any others after that are really only for tie-breaking if the first one is a 5. There may be some extreme edgecases where you've got something like ".50000000000000009" where rounding twice in this way not going to be perfect, but I'm willing to accept that trade-off to avoid the much more common and significant rounding errors that would happen in the other example.

gregtatcam commented on src/libxrpl/protocol/STNumber.cpp:
```diff
@@ -179,20 +243,24 @@ numberFromJson(SField const& field, Json::Value const& value)
-            Throw<std::range_error>("too high");
+        // Number mantissas are much bigger than the allowable parsed values, so
+        // it can't be out of range.
+        static_assert(
```
Shouldn't this change be amendment gated since the exception is no longer thrown?

ximinez replied:
> Shouldn't this change be amendment gated since the exception is no longer thrown?

No, I don't think so, because this is RPC functionality.

The only callers are in `STParsedJSON` and unit tests.

Instead, I'll assert `!getCurrentTransactionRules()`, which will indicate we're not doing transaction processing.

4605bbfb6f1e1464fb15b076a3ffcd7efe68d916

gregtatcam commented on src/xrpld/app/paths/detail/AMMLiquidity.cpp:
```diff
@@ -42,9 +42,11 @@ AMMLiquidity<TIn, TOut>::generateFibSeqOffer(
 {
     TAmounts<TIn, TOut> cur{};
 
+    Number const initialFibSeqPct =
```
Why is this copy needed instead of directly multiplying by `InitialFibSeqPct`?

ximinez replied:
> Why is this copy needed instead of directly multiplying by `InitialFibSeqPct`?

I did it so that the value would be re-normalized before being used.

Now that I look at it again, the only reason that re-normalizing was needed is because `operator*=` has an assertion that the parameters are normalized. I removed that assertion, because the _result_ will be normalized, regardless of how big the original mantissas are.

4605bbfb6f1e1464fb15b076a3ffcd7efe68d916

gregtatcam commented on src/libxrpl/protocol/STTakesAsset.cpp:
```diff
@@ -0,0 +1,35 @@
+}
+
+void
+associateAsset(SLE& sle, Asset const& asset)
```
Nothing comes to mind but what if multiple assets have to be associated with a ledger object?

ximinez replied:
> Nothing comes to mind but what if multiple assets have to be associated with a ledger object?

Ummmmm, I guess we'll deal with that situation when we come to it.

ximinez commented:
> Want to make sure I understood this correctly: with largeRange, we can support full precision up to 2^63 - 1, so integer types (MPT or XRP) don’t lose precision even for large values. We’re still bounded by the signed `int64` limit, but this works for existing integers since none of them exceed `int64`.

That is correct.

> However, if we ever want to go beyond the 63-bit range (e.g., increasing MPT’s `MaximumAmount` to `max(uint64)` in the future), this implementation is not sufficient.

That is correct, however my current understanding is that there is no need or plan to extend the current limit.

Such a change would require an amendment, so a second expansion of `Number` would need to be included in that amendment work.

(Expounded in this comment: https://github.com/XRPLF/rippled/pull/6025#issuecomment-3720510150)

gregtatcam commented on src/xrpld/app/tx/detail/LoanPay.cpp:
```diff
@@ -430,28 +431,30 @@ LoanPay::doApply()
+    {
+        // LCOV_EXCL_START
+        [[maybe_unused]]
+        auto const available = *assetsAvailableProxy;
```
Why are `available` and `total` needed? They are not used.

ximinez replied:
> Why are `available` and `total` needed? They are not used.

This whole block has since been replaced.

ximinez replied:
> Why are `available` and `total` needed? They are not used.

This whole block has since been replaced.

gregtatcam commented on src/xrpld/app/tx/detail/LoanBrokerCoverClawback.cpp:
```diff
@@ -330,6 +332,8 @@ LoanBrokerCoverClawback::doApply()
     sleBroker->at(sfCoverAvailable) -= clawAmount;
     view().update(sleBroker);
 
+    associateAsset(*sleBroker, vaultAsset);
```
I keep on thinking if there is a way to make this association table driven. There is already a new field property defined `sMD_NeedsAsset`. What if instead adding this property in `sfields`, it's added in `ledger_entries`. Then we just need a way to point it to the associated asset, which requires accessing up to two linked objects right now. What if we add an integral flag to each related ledger entry (Vault, LoanBroker, Loan) to indicate if an asset is integral? Then in `STObject::applyTemplate()`, which iterates over all fields , for any `STNumber` field with `sMD_NeedsAsset`, we can set the integral flag. Does it make sense to explore this approach?

ximinez replied:
> I keep on thinking if there is a way to make this association table driven. There is already a new field property defined `sMD_NeedsAsset`. What if instead adding this property in `sfields`, it's added in `ledger_entries`. Then we just need a way to point it to the associated asset, which requires accessing up to two linked objects right now. What if we add an integral flag to each related ledger entry (Vault, LoanBroker, Loan) to indicate if an asset is integral? Then in `STObject::applyTemplate()`, which iterates over all fields , for any `STNumber` field with `sMD_NeedsAsset`, we can set the integral flag. Does it make sense to explore this approach?

Well, `sMD_NeedsAsset` is an `SField`-only property, so it would have to be set there. `SOEStyle` does not use bit flags, and isn't as flexible.

That aside, what I think you're suggesting is define a new `LedgerSpecificFlags` called `lsfIntegralAsset` or something, that could be used with Vault, LoanBroker, and Loan. Instead of needing to pull the `Vault.Asset` in every transaction, we would set that flag on creation of the Vault, and then read and set the flag when creating the derived objects.

Then instead of `STTakesAsset`, we'd check the `Flags` field in `STObject::applyTemplate()`, and if that flag is set, also set an `isIntegral` flag in every `STNumber` we come across that has the `sMD_NeedsAsset` property.

If the flag is set, `STNumber` would round to integer when serializing, and we're done.

It's a neat idea, and worth exploring. There are a couple of potential problems I can think of offhand:
1. I don't think the flag could be global, because it would be difficult to keep it unique. Ledger flags doesn't have a concept of "universal flags" like transactions do. The reason that's significant is that `applyTemplate` would then have to understand
	1. that the object is a ledger object
	2. which type of ledger object it is
	3. which flag corresponds to "integral" for that ledger object type.
2. I'm not entirely sure where in the lifecycle of an object `applyTemplate` is called. I think it's toward the beginning - when the object is created or read. Problem with that, as I just documented in 4605bbfb6f1e1464fb15b076a3ffcd7efe68d916 is that setting the `STNumber` flag would have to be done _after_ modifications are made and fields are set, because there's no `STNumber` to work with when the field is missing/unset. There may be a better call point just before serializing, so this isn't an insurmountable problem. (Maybe `STObject::add`.)
3. It might take too long to implement and review this before we want to get this PR merged, and the amendments are enabled. It could be added later with another amendment, but then you have backwards compatibility to deal with. Unless we define the flags, and set them in their respective creation transactions, _before_ enabling the amendments, then come back and update the implementation later... That seems hinky, though.

gregtatcam replied:
Yes, maybe can think about it as an enhancement after the Lending Protocol is released. This is not going to require an amendment.

gregtatcam reviewed (commented):
Looks good to me. Thanks for adding extensive comments to the Number and MantissaRange classes.

Please add some documentation to STTakesAsset.h explaining why it's needed and how to use it. I'll approve once those are in!

ximinez commented:
> > Want to make sure I understood this correctly: with largeRange, we can support full precision up to 2^63 - 1, so integer types (MPT or XRP) don’t lose precision even for large values. We’re still bounded by the signed `int64` limit, but this works for existing integers since none of them exceed `int64`.
> 
> That is correct.
> 
> > However, if we ever want to go beyond the 63-bit range (e.g., increasing MPT’s `MaximumAmount` to `max(uint64)` in the future), this implementation is not sufficient.
> 
> That is correct, however my current understanding is that there is no need or plan to extend the current limit.
> 
> Such a change would require an amendment, so a second expansion of `Number` would need to be included in that amendment work.

The updated class documentation now [explicitly mentions that the limits for `Number` were chosen partially because of `maxMPTokenAmount`](https://github.com/XRPLF/rippled/blob/c5c89c28a76a1da385c75701f3ec2e031b4d4d03/include/xrpl/basics/Number.h#L156-L162).

I've added a couple of `static_assert`s to ensure this invariant remains true in c88dd55d07

gregtatcam reviewed (approved):
LGTM

ximinez committed:
Expand Number to support the full integer range (#6025)

- Refactor Number internals away from int64 to uint64 & a sign flag
  - ctors and accessors return `rep`. Very few things expose
    `internalrep`.
  - An exception is "unchecked" and the new "normalized", which explicitly
    take an internalrep. But with those special control flags, it's easier
    to distinguish and control when they are used.

- For now, skip the larger mantissas in AMM transactions and tests

- Remove trailing zeros from scientific notation Number strings
  - Update tests. This has the happy side effect of making some of the string
    representations _more_ consistent between the small and large
    mantissa ranges.

- Add semi-automatic rounding of STNumbers based on Asset types
  - Create a new SField metadata enum, sMD_NeedsAsset, which indicates
    the field should be associated with an Asset so it can be rounded.
  - Add a new STTakesAsset intermediate class to handle the Asset
    association to a derived ST class. Currently only used in STNumber,
    but could be used by other types in the future.
  - Add "associateAsset" which takes an SLE and an Asset, finds the
    sMD_NeedsAsset fields, and associates the Asset to them. In the case
    of STNumber, that both stores the Asset, and rounds the value
    immediately.
  - Transactors only need to add a call to associateAsset _after_ all of
    the STNumbers have been set. Unfortunately, the inner workings of
    STObject do not do the association correctly with uninitialized
    fields.
  - When serializing an STNumber that has an Asset, round it before
    serializing.
  - Add an override of roundToAsset, which rounds a Number value in place
    to an Asset, but without any additional scale.
  - Update and fix a bunch of Loan-related tests to accommodate the
    expanded Number class.

---------

Co-authored-by: Vito <5780819+Tapanito@users.noreply.github.com>

ximinez commented:
As with the parent PR (#6102), this branch had missing signatures preventing merge. I squashed the commits down to one and pushed.

> [ximinez](https://github.com/ximinez) [force-pushed](https://github.com/XRPLF/rippled/compare/4ec25273958ab58e6ba2492fad39a3599ed1cbd7..59b07d515a8cf1ff670a209a561241ee687315b3) the ximinez/lending-number-simple branch from [4ec2527](https://github.com/XRPLF/rippled/commit/4ec25273958ab58e6ba2492fad39a3599ed1cbd7) to [59b07d5](https://github.com/XRPLF/rippled/commit/59b07d515a8cf1ff670a209a561241ee687315b3)

```
$ git diff 4ec2527 59b07d5 

$ git diff 4ec2527 59b07d5 | wc
      0       0       0
```
